File: java/src/main/java/ai/buble/sdk/BubleClient.java
package ai.buble.sdk;

import ai.buble.sdk.apps.AppsService;
import ai.buble.sdk.chat.ChatService;
import ai.buble.sdk.files.FilesService;
import ai.buble.sdk.generations.GenerationsService;
import ai.buble.sdk.http.BubleHttpClient;
import ai.buble.sdk.media.MediaModelsService;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;

import java.net.http.HttpClient;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Server-side client for the Buble public API.
 */
public final class BubleClient {
    public static final String DEFAULT_BASE_URL = "https://buble.ai";
    public static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(60);

    private final BubleHttpClient http;
    private final MediaModelsService mediaModels;
    private final FilesService files;
    private final GenerationsService generations;
    private final AppsService apps;
    private final ChatService chat;

    private BubleClient(Builder builder) {
        String resolvedApiKey = firstNonEmpty(builder.apiKey, System.getenv("BUBLE_API_KEY"));
        String resolvedBaseUrl = firstNonEmpty(builder.baseUrl, System.getenv("BUBLE_BASE_URL"), DEFAULT_BASE_URL);
        if (resolvedApiKey == null || resolvedApiKey.isEmpty()) {
            throw new BubleException("Missing Buble API key. Pass apiKey or set BUBLE_API_KEY.");
        }
        ObjectMapper mapper = builder.objectMapper == null ? defaultObjectMapper() : builder.objectMapper;
        this.http = new BubleHttpClient(
                resolvedApiKey,
                resolvedBaseUrl,
                builder.timeout,
                builder.httpClient,
                mapper,
                builder.headers
        );
        this.mediaModels = new MediaModelsService(http);
        this.files = new FilesService(http);
        this.generations = new GenerationsService(http);
        this.apps = new AppsService(http);
        this.chat = new ChatService(http);
    }

    public static BubleClient fromEnv() {
        return builder().build();
    }

    public static Builder builder() {
        return new Builder();
    }

    public String getBaseUrl() {
        return http.getBaseUrl();
    }

    public MediaModelsService mediaModels() {
        return mediaModels;
    }

    public FilesService files() {
        return files;
    }

    public GenerationsService generations() {
        return generations;
    }

    public AppsService apps() {
        return apps;
    }

    public ChatService chat() {
        return chat;
    }

    private static ObjectMapper defaultObjectMapper() {
        ObjectMapper mapper = new ObjectMapper();
        mapper.setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE);
        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
        return mapper;
    }

    private static String firstNonEmpty(String... values) {
        for (String value : values) {
            if (value != null && !value.isEmpty()) {
                return value;
            }
        }
        return null;
    }

    public static final class Builder {
        private String apiKey;
        private String baseUrl;
        private Duration timeout = DEFAULT_TIMEOUT;
        private HttpClient httpClient = HttpClient.newHttpClient();
        private ObjectMapper objectMapper;
        private final Map<String, String> headers = new LinkedHashMap<String, String>();

        public Builder apiKey(String apiKey) {
            this.apiKey = apiKey;
            return this;
        }

        public Builder baseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        public Builder timeout(Duration timeout) {
            if (timeout != null) {
                this.timeout = timeout;
            }
            return this;
        }

        public Builder httpClient(HttpClient httpClient) {
            if (httpClient != null) {
                this.httpClient = httpClient;
            }
            return this;
        }

        public Builder objectMapper(ObjectMapper objectMapper) {
            this.objectMapper = objectMapper;
            return this;
        }

        public Builder header(String key, String value) {
            if (key != null && value != null) {
                this.headers.put(key, value);
            }
            return this;
        }

        public BubleClient build() {
            return new BubleClient(this);
        }
    }
}

File: swift/Sources/Buble/BubleClient.swift
/// Server-side client for the Buble public API.
public final class BubleClient: @unchecked Sendable {
    public let mediaModels: MediaModelsService
    public let files: FilesService
    public let generations: GenerationsService
    public let apps: AppsService
    public let chat: ChatService

    let http: BubleHTTPClient

    public init(options: BubleClientOptions) throws {
        let apiKey = options.apiKey ?? ProcessInfo.processInfo.environment["BUBLE_API_KEY"]
        guard let apiKey, !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw BubleError.missingAPIKey
        }

        let baseURL: URL
        if let configuredBaseURL = options.baseURL {
            baseURL = configuredBaseURL
        } else if let value = ProcessInfo.processInfo.environment["BUBLE_BASE_URL"], !value.isEmpty {
            guard let url = URL(string: value) else {
                throw BubleError.invalidURL(value)
            }
            baseURL = url
        } else {
            baseURL = BubleClientOptions.defaultBaseURL
        }

        let http = BubleHTTPClient(
            apiKey: apiKey,
            baseURL: baseURL,
            timeout: options.timeout,
            headers: options.headers,
            transport: options.transport ?? URLSessionHTTPTransport()
        )

        self.http = http
        self.mediaModels = MediaModelsService(http: http)
        self.files = FilesService(http: http)
        self.generations = GenerationsService(http: http)
        self.apps = AppsService(http: http)
        self.chat = ChatService(http: http)
    }

    public convenience init(apiKey: String) throws {
        try self.init(options: BubleClientOptions(apiKey: apiKey))
    }

    /// Creates a client from `BUBLE_API_KEY` and optional `BUBLE_BASE_URL`.
    public static func fromEnvironment() throws -> BubleClient {
        try BubleClient(options: BubleClientOptions())
    }
}

/// Namespace marker for the Buble Swift SDK.
public enum Buble {}
